// react-native-css-interop/dist/shared.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRuntimeValue = exports.isPropDescriptor = exports.STYLE_SCOPES = exports.DEFAULT_CONTAINER_NAME = exports.INTERNAL_FLAGS = exports.INTERNAL_SET = exports.INTERNAL_RESET = void 0;
exports.INTERNAL_RESET = Symbol();
exports.INTERNAL_SET = Symbol();
exports.INTERNAL_FLAGS = Symbol();
exports.DEFAULT_CONTAINER_NAME = "@__";
exports.STYLE_SCOPES = {
    GLOBAL: 0,
    CONTEXT: 1,
    SELF: 2,
};
function isPropDescriptor(value) {
    return typeof value === "object" && value !== null && "$$type" in value;
}
exports.isPropDescriptor = isPropDescriptor;
function isRuntimeValue(value) {
    if (!value) {
        return false;
    }
    else if (Array.isArray(value)) {
        return value.some((v) => isRuntimeValue(v));
    }
    else if (typeof value === "object") {
        if (value.type === "runtime") {
            return true;
        }
        else {
            return Object.values(value).some((v) => isRuntimeValue(v));
        }
    }
    else {
        return false;
    }
}
exports.isRuntimeValue = isRuntimeValue;
