// react-native-css-interop/dist/css-to-rn/functions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.platformColor = exports.roundToNearestPixel = exports.getPixelSizeForLayoutSize = exports.fontScale = exports.pixelScale = exports.fontScaleSelect = exports.pixelScaleSelect = exports.platformSelect = exports.hairlineWidth = void 0;
function hairlineWidth() {
    return "hairlineWidth()";
}
exports.hairlineWidth = hairlineWidth;
function platformSelect(specifics) {
    let output = [];
    for (const [key, value] of Object.entries(specifics)) {
        output.push(`${key}/${value}`);
    }
    return `platformSelect(${output.join(",")})`;
}
exports.platformSelect = platformSelect;
function pixelScaleSelect(specifics) {
    let output = [];
    for (const [key, value] of Object.entries(specifics)) {
        output.push(`${key}/${value}`);
    }
    return `pixelScaleSelect(${output.join(",")})`;
}
exports.pixelScaleSelect = pixelScaleSelect;
function fontScaleSelect(specifics) {
    let output = [];
    for (const [key, value] of Object.entries(specifics)) {
        output.push(`${key}/${value}`);
    }
    return `fontScaleSelect(${output.join(",")})`;
}
exports.fontScaleSelect = fontScaleSelect;
function pixelScale(value) {
    return value === undefined ? `pixelScale()` : `pixelScale(${value})`;
}
exports.pixelScale = pixelScale;
function fontScale(value) {
    return value === undefined ? `fontScale()` : `fontScale(${value})`;
}
exports.fontScale = fontScale;
function getPixelSizeForLayoutSize(value) {
    return `getPixelSizeForLayoutSize(${value})`;
}
exports.getPixelSizeForLayoutSize = getPixelSizeForLayoutSize;
function roundToNearestPixel(value) {
    return `roundToNearestPixel(${value})`;
}
exports.roundToNearestPixel = roundToNearestPixel;
function platformColor(...colors) {
    return `platformColor(${colors
        .map((color) => {
        return color
            .replaceAll("?", "\\?")
            .replaceAll("/", "\\/")
            .replaceAll(":", "\\:");
    })
        .join(",")})`;
}
exports.platformColor = platformColor;
