// react-native-css-interop/dist/css-to-rn/functions-web.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.platformColor = exports.roundToNearestPixel = exports.getPixelSizeForLayoutSize = exports.fontScale = exports.pixelScale = exports.fontScaleSelect = exports.pixelScaleSelect = exports.platformSelect = exports.hairlineWidth = void 0;
function hairlineWidth() {
    return 1;
}
exports.hairlineWidth = hairlineWidth;
function platformSelect(specifics) {
    return specifics["web"] ?? specifics["default"];
}
exports.platformSelect = platformSelect;
function pixelScaleSelect(specifics) {
    return specifics[1] ?? specifics["default"];
}
exports.pixelScaleSelect = pixelScaleSelect;
function fontScaleSelect(specifics) {
    return specifics[1] ?? specifics["default"];
}
exports.fontScaleSelect = fontScaleSelect;
function pixelScale(value = 1) {
    return value;
}
exports.pixelScale = pixelScale;
function fontScale(value = 1) {
    return value;
}
exports.fontScale = fontScale;
function getPixelSizeForLayoutSize(value) {
    return value;
}
exports.getPixelSizeForLayoutSize = getPixelSizeForLayoutSize;
function roundToNearestPixel(value) {
    return value;
}
exports.roundToNearestPixel = roundToNearestPixel;
function platformColor(...colors) {
    return colors;
}
exports.platformColor = platformColor;
