// react-native-css-interop/dist/testing-library/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.revealStyles = exports.registerCSS = exports.resetComponents = exports.resetStyles = exports.createRemappedComponent = exports.createMockComponent = exports.warnings = void 0;
const react_1 = require("react");
const JSX = __importStar(require("react/jsx-runtime"));
const index_1 = require("../index");
const shared_1 = require("../shared");
const css_to_rn_1 = require("../css-to-rn");
require("../runtime/components");
const rendering_1 = require("../runtime/components/rendering");
const globals_1 = require("../runtime/native/globals");
__exportStar(require("../types"), exports);
var globals_2 = require("../runtime/native/globals");
Object.defineProperty(exports, "warnings", { enumerable: true, get: function () { return globals_2.warnings; } });
function createMockComponent(Component, mapping = {
    className: "style",
}) {
    (0, rendering_1.cssInterop)(Component, mapping);
    const mock = jest.fn(({ ...props }, ref) => {
        props.ref = ref;
        return (0, rendering_1.render)(JSX.jsx, Component, props, "");
    });
    const component = (0, react_1.forwardRef)(mock);
    return Object.assign(component, { mock });
}
exports.createMockComponent = createMockComponent;
function createRemappedComponent(Component, mapping) {
    (0, rendering_1.remapProps)(Component, mapping);
    const mock = jest.fn((props, ref) => {
        props.ref = ref;
        return (0, rendering_1.render)(JSX.jsx, Component, props, "");
    });
    const component = (0, react_1.forwardRef)(mock);
    return Object.assign(component, { mock });
}
exports.createRemappedComponent = createRemappedComponent;
const resetStyles = () => {
    index_1.StyleSheet[shared_1.INTERNAL_RESET]();
};
exports.resetStyles = resetStyles;
const resetComponents = () => {
    rendering_1.interopComponents.clear();
};
exports.resetComponents = resetComponents;
function registerCSS(css, options) {
    index_1.StyleSheet.registerCompiled((0, css_to_rn_1.cssToReactNativeRuntime)(css, options));
}
exports.registerCSS = registerCSS;
function revealStyles(obj) {
    switch (typeof obj) {
        case "string":
        case "number":
        case "bigint":
        case "boolean":
        case "symbol":
        case "undefined":
        case "function":
            return obj;
        case "object":
        default: {
            const style = globals_1.opaqueStyles.get(obj);
            if (style)
                return style;
            return Object.fromEntries(Object.entries(obj).map(([key, value]) => {
                switch (typeof value) {
                    case "string":
                    case "number":
                    case "bigint":
                    case "boolean":
                    case "symbol":
                    case "undefined":
                    case "function":
                        return [key, value];
                    case "object":
                    default: {
                        if (Array.isArray(value)) {
                            return [key, value.map(revealStyles)];
                        }
                        else if (value) {
                            const style = globals_1.opaqueStyles.get(value);
                            return [key, style ?? value];
                        }
                        else {
                            return [key, value];
                        }
                    }
                }
            }));
        }
    }
}
exports.revealStyles = revealStyles;
