// babel-plugin-tester/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  pluginTester: true,
  prettierFormatter: true,
  unstringSnapshotSerializer: true,
  runPluginUnderTestHere: true,
  runPresetUnderTestHere: true
};
exports.pluginTester = exports.default = defaultPluginTester;
Object.defineProperty(exports, "prettierFormatter", {
  enumerable: true,
  get: function () {
    return _prettier.prettierFormatter;
  }
});
Object.defineProperty(exports, "runPluginUnderTestHere", {
  enumerable: true,
  get: function () {
    return _pluginTester.runPluginUnderTestHere;
  }
});
Object.defineProperty(exports, "runPresetUnderTestHere", {
  enumerable: true,
  get: function () {
    return _pluginTester.runPresetUnderTestHere;
  }
});
Object.defineProperty(exports, "unstringSnapshotSerializer", {
  enumerable: true,
  get: function () {
    return _unstringSnapshot.unstringSnapshotSerializer;
  }
});
var _debug = _interopRequireDefault(require("debug"));
var _prettier = require("./formatters/prettier");
var _unstringSnapshot = require("./serializers/unstring-snapshot");
var _pluginTester = require("./plugin-tester");
var _types = require("./types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
var _expect;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const debug = (0, _debug.default)('babel-plugin-tester:index');
if ('expect' in globalThis && typeof ((_expect = expect) === null || _expect === void 0 ? void 0 : _expect.addSnapshotSerializer) == 'function') {
  debug('added unstring snapshot serializer globally; all snapshots after this point will be affected');
  expect.addSnapshotSerializer(_unstringSnapshot.unstringSnapshotSerializer);
} else {
  debug('unable to add unstring snapshot serializer: global expect object is missing or unsupported');
}
function defaultPluginTester(options) {
  return (0, _pluginTester.pluginTester)({
    formatResult: _prettier.prettierFormatter,
    ...options
  });
}
defaultPluginTester.default = defaultPluginTester;
defaultPluginTester.pluginTester = defaultPluginTester;
defaultPluginTester.prettierFormatter = _prettier.prettierFormatter;
defaultPluginTester.unstringSnapshotSerializer = _unstringSnapshot.unstringSnapshotSerializer;
defaultPluginTester.runPluginUnderTestHere = _pluginTester.runPluginUnderTestHere;
defaultPluginTester.runPresetUnderTestHere = _pluginTester.runPresetUnderTestHere;
module.exports = defaultPluginTester;
